### PR TITLE
[networks] Add prototype for HTTP monitoring in eBPF 

### PR DESCRIPTION
**Pull request by p-lambert**
#### Overview

This PR provides a HTTP monitoring prototype using eBPF. In a nutshell, we track HTTP transactions by attaching an eBPF `BPF_PROG_TYPE_SOCKET_FILTER` program to a “headless” `SOCK_RAW` and then inspecting traffic hitting ports 80 and 8080.

From each ethernet frame matching this criteria we extract a `conn_tuple_t` similarly to what we do on the `kprobe` programs.

Then we do a map lookup in order to retrieve the associated `http_transaction_t`. This embeds information such as the state of each transaction (eg. whether we’re processing a request or response), the associated HTTP method, the status code of the response, the latency of the whole transaction as well as a fragment of the request. 

The state machine governing the processing is quite simple:
```
HTTP_UNKNOWN -> HTTP_REQUESTING -> HTTP_RESPONDING
``` 

When there is no state associated with the current TCP connection (`HTTP_UNKNOWN`) we read up to `HTTP_BUFFER_SIZE` bytes from the TCP segment.

If we match `^(GET|POST|PUT|DELETE|HEAD)` we copy the request fragment to `http_transaction`  along with the time of the request and transition to the `HTTP_REQUESTING` state.

For each following segment we inspect again the first `HTTP_BUFFER_SIZE` bytes of the payload until we match `^HTTP` (note that an HTTP request can be split across multiple segments). Once that happens we parse the status code and transition to `HTTP_RESPONDING`.

For each subsequent segment we update the `response_last_seen` timestamp field from the `http_transaction_t` until one of the two conditions happen:
* A TCP segment with a `FIN` flag is detected;
* A new request is detected, that is, we see another TCP segment beginning with `^(GET|POST|PUT|DELETE|HEAD)`  (which can happen in the context of HTTP keep-alives);

Once that happens we consider this HTTP transaction “done” and enqueue the `http_transaction_t` struct in another eBPF map where transactions are batched for consumption from userspace. 

#### Batching

**UPDATE** I'm trying a different approach in https://github.com/DataDog/datadog-agent/pull/6740/files

From our past experience with TCP closed connections (which are flushed using perf events), we know that we need some sort of batching to reduce the number of times we poll the perf ring in order to keep the overhead of system-probe low in high throughput workloads.

Here we take a similar approach and using a `http_batch_t` struct that looks like the following:

batch

Basically each `http_batch_t` entry is a circular buffer divided in multiple “pages”. Once we fill up a whole page with `http_transaction_t` structs, we send a notification `http_notification_t` perf event to the userspace program, which essentially conveys that “page X from batch from CPU Y” can be read. 

There are obviously issues with this approach and I’ll try to explain some of the tradeoffs involved in the following sections. 

Finally there are two params that can be configured regarding the batching mechanism:
* `HTTP_BATCH_SIZE` : determines the size of each “page”. The larger this number is the lower the CPU overhead of consuming the perf events. The downside is increased latency.
* `HTTP_BATCH_PAGES`: determines the number of pages for each batch object. In theory the greater this number is, the less likely is missing events due to a data race. (Having more pages gives more time to the userspace program to read the page before it gets overridden);

#### Why not use TC classifiers?

I haven’t fully thought about this yet, but the main reason we’re not using a TC program here is because you need to attach it to a specific device, and that would be tricky in the context of ephemeral containers.
However one major downside of socket filter programs compared to TC classifiers is that you don’t have access to the `bpf_perf_event_output` helper until [Linux 5.4](https://github.com/torvalds/linux/commit/7c4b90d79d0f4ee6f2c01a114ef0a83a09dfc900). Because of that we have to resort to a hack of flushing the perf events from a `kprobe`. Here we’re using `kretprobe/tcp_sendmsg`  (where `http_notify_batch` is called) but I haven’t thought much about that yet, so there are likely better hook points for that.

In any case most of the code would remain the same regardless of the eBPF program type we decide to use.

#### Why not flush the batch of HTTP transactions directly to the perf ring?

TL;DR to support Kernels <4.11.
The reason why we send a `http_notification_t` object to userspace as opposed to a batch of transactions is because you can’t flush map entries directly to the perf ring until Kernel 4.11. Refer to [this JIRA card](https://datadoghq.atlassian.net/browse/NET-869?focusedCommentId=294037) for the code bits showing this.

Without being able to do that, we would need to copy the batch into the eBPF stack which is limited to 512bytes. I tried to that and we blow up the stack with a batch size of 2.

On the other hand that means that if we drop support for Kernel versions older than 4.11 we can simplify the batching logic by a lot: we would get rid of the map lookup from userspace, the `http_notification_t` bits, and wouldn’t need the multi page stuff as we wouldn’t need a circular buffer. 

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### CODE REVIEW

File: pkg/ebpf/c/bpf_helpers.h [hunk 1]
```diff
@@ -52,6 +52,11 @@ static int (*bpf_skb_set_tunnel_key)(void* ctx, void* key, int size, int flags)
 static unsigned long long (*bpf_get_prandom_u32)(void) = (void*)BPF_FUNC_get_prandom_u32;
 static u64 (*bpf_get_current_task)(void) = (void *) BPF_FUNC_get_current_task;
 
+static inline __attribute__((always_inline))
+u32 bpf_ntohl(u32 val) {
+  return __builtin_bswap32(val);
```
Comment (brycekahle) [inline pkg/ebpf/c/bpf_helpers.h:57]: Note: This is only correct for little endian systems. See https://github.com/torvalds/linux/blob/47ec5303d73ea344e84f46660fff693c57641386/tools/lib/bpf/bpf_endian.h#L50-L78

File: pkg/ebpf/c/tracer-ebpf.c [hunk 2]
```diff
@@ -1131,52 +1185,285 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
...
+    // Among other things, but nothing really worked on Kernel 4.4
+    // It seems that indeed support for this type of access by the verifier was added later on:
+    // https://patchwork.ozlabs.org/project/netdev/patch/1475074472-23538-1-git-send-email-jbacik@fb.com/
+    //
+    // What is unfortunate about this is not only that enqueing a HTTP transaction is O(HTTP_BATCH_SIZE*HTTP_BATCH_PAGES),
+    // but also that we can't really increase the batch/page size at the moment because that blows up the eBPF *program* size
+#pragma unroll
+    for (int i = 0; i < HTTP_BATCH_PAGES*HTTP_BATCH_SIZE; i++) {
+        if (i == batch_slot) {
+            __builtin_memcpy(&batch->txs[i], http, sizeof(http_transaction_t));
```
Comment (leeavital) [inline pkg/ebpf/c/tracer-ebpf.c:1320]: I wonder if we could use one of the atomic intrinsics to avoid writing to txs[i] while userspace reads it. 

So userspace could write FREE to txs[i] and the kprobe could do  

```
if ! ( compare_and_swap(txs[i], FREED, WRITING) ) {
    // increment counter for dropped HTTP event
} else {
   __builtin_memcpy(&batch->txs[i], http, sizeof(http_transaction_t));
}
```

File: pkg/ebpf/c/tracer-ebpf.c [hunk 3]
```diff
@@ -1131,52 +1185,285 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
...
+            response_code = (buffer[i+1]-'0')*100;
+        }
+    }
+
+    if (response_code < 100 || response_code >= 600) {
+        return 0;
+    }
+
+    http->state = HTTP_RESPONDING;
+    http->response_code = response_code;
```
Comment (brycekahle) [inline pkg/ebpf/c/tracer-ebpf.c:1365]: If the intention is to only capture the "class" of status code, then we should be careful to label it appropriately. I could see calling it `response_code` being misleading, if it was only 1xx/2xx/3xx/4xx/5xx.

File: pkg/ebpf/c/tracer-ebpf.c [hunk 4]
```diff
@@ -1131,52 +1185,285 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
...
+        packet_type = HTTP_REQUESTING_GET;
+    } else if ((p[0] == 'P') && (p[1] == 'O') && (p[2] == 'S') && (p[3] == 'T')) {
+        packet_type = HTTP_REQUESTING_POST;
+    } else if ((p[0] == 'P') && (p[1] == 'U') && (p[2] == 'T')) {
+        packet_type = HTTP_REQUESTING_PUT;
+    } else if ((p[0] == 'D') && (p[1] == 'E') && (p[2] == 'L') && (p[3] == 'E') && (p[4] == 'T') && (p[5] == 'E')) {
+        packet_type = HTTP_REQUESTING_DELETE;
+    } else if ((p[0] == 'H') && (p[1] == 'E') && (p[2] == 'A') && (p[3] == 'D')) {
+        packet_type = HTTP_REQUESTING_HEAD;
+    }
```
Comment (brycekahle) [inline pkg/ebpf/c/tracer-ebpf.c:1396]: Are we planning to support the other common methods? `CONNECT`, `OPTIONS`, and `PATCH`?

File: pkg/ebpf/c/tracer-ebpf.c [hunk 5]
```diff
@@ -1131,52 +1185,285 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
...
+    // Find digit following space and multiply it by 100 to obtain the response code
+    // HTTP/1.1 200 OK
+    // _________^_____
+    __u16 response_code = 0;
+#pragma unroll
+    for (int i = 0; i < HTTP_BUFFER_SIZE-1; i++) {
+        if (response_code == 0 && buffer[i] == ' ') {
+            response_code = (buffer[i+1]-'0')*100;
+        }
+    }
```
Comment (leeavital) [inline pkg/ebpf/c/tracer-ebpf.c:1358]: Will this pick up error codes like 404 ? It looks like it only takes the first digit into account

File: pkg/ebpf/c/tracer-ebpf.c (hunk 2 above)
Comment (leeavital) [reply pkg/ebpf/c/tracer-ebpf.c:1320]: nevermind, this relies on being able to write to part of the batch from userspace which isn't possible

File: pkg/network/socket_filter.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+	nsErr := util.WithRootNS(rootPath, func() {
+		packetSrc, srcErr = newPacketSource(filter)
+	})
+	if nsErr != nil {
+		return nil, nsErr
+	}
+	if srcErr != nil {
+		return nil, srcErr
+	}
```
Comment (sunhay) [inline pkg/network/socket_filter.go:28]: Besides the point of this PR, but this is a bit awkward  to read - we could probably just change the function argument interface so its written as:

```go
err := util.WithRootNS(rootPath, func() err {
	packetSrc, srcErr = newPacketSource(filter)
	return srcErr
})
if err != nil {
	return nil, err
}
```

where `err = nsErr or srcErr` within `util.WithRootNS()`

File: pkg/ebpf/c/tracer-ebpf.h [hunk 7]
```diff
@@ -56,6 +56,72 @@ typedef struct {
...
+
+typedef enum {
+    HTTP_UNKNOWN           = 0,
+    HTTP_RESPONDING        = 1,
+    HTTP_REQUESTING_GET    = 1 << 1,
+    HTTP_REQUESTING_POST   = 1 << 2,
+    HTTP_REQUESTING_PUT    = 1 << 3,
+    HTTP_REQUESTING_DELETE = 1 << 4,
+    HTTP_REQUESTING_HEAD   = 1 << 5,
+} http_state_t;
```
Comment (brycekahle) [inline pkg/ebpf/c/tracer-ebpf.h:86]: This feels confusing, as it is mixing two concerns:
1. the state of the transaction: unknown->requesting->responding
2. The request method

It is also referenced in code as both `state` and `packet_type`. Perhaps they should be separate things, or at least different masked parts of the byte.

Method also doesn't need to be a bitmask, since it cannot be multiple methods at the same time.

File: pkg/ebpf/http.go [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+func toHTTPNotification(data []byte) httpNotification {
+	return *(*httpNotification)(unsafe.Pointer(&data[0]))
+}
+
+// Path returns the URL from the request fragment captured in eBPF
+// Usually the request fragment will look like
+// GET /foo HTTP/1.1
+func (tx *httpTX) Path() string {
+	b := C.GoBytes(unsafe.Pointer(&tx.request_fragment), C.int(C.HTTP_BUFFER_SIZE))
```
Comment (brycekahle) [inline pkg/ebpf/http.go:39]: Since it seems you concerned about performance of this function (due to the loops below, rather than something like `strings.Split`), if you aren't aware,  `C.GoBytes` will copy (and thus allocate). If you want to iterate on the raw bytes, you can instead do `(*[C.HTTP_BUFFER_SIZE]byte)(unsafe.Pointer(&tx.request_fragment))` and then only do a `copy` once you've found the range you are interested in.

You may also want to cache that return value with a `sync.Once` or something, if this function may be called a lot.

File: pkg/ebpf/http.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+		var (
+			misses int
+			then   = time.Now()
+			hits   = make(map[int]int)
+			report = time.NewTicker(30 * time.Second)
+		)
+
+		for {
+			select {
+			case data, ok := <-http.perfHandler.ClosedChannel:
```
Comment (brycekahle) [inline pkg/ebpf/http.go:137]: We probably want to rename `ClosedChannel` to something more generic, since Closed doesn't make much sense in this context.

File: pkg/ebpf/http.go [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+				}
+			case _, ok := <-http.perfHandler.LostChannel:
+				if !ok {
+					return
+				}
+				misses++
+			case now := <-report.C:
+				delta := float64(now.Sub(then).Seconds())
+				log.Infof("http report: 200(%d reqs, %.2f/s) 300(%d reqs, %.2f/s), 400(%d reqs, %.2f/s) 500(%d reqs, %.2f/s), misses(%d reqs, %.2f/s)",
+					hits[200], float64(hits[200])/delta,
```
Comment (brycekahle) [inline pkg/ebpf/http.go:171]: Should this include `hits[100]`?

File: pkg/ebpf/c/bpf_helpers.h (hunk 1 above)
Comment (p-lambert) [reply pkg/ebpf/c/bpf_helpers.h:57]: Good catch 👍

File: pkg/ebpf/c/tracer-ebpf.c (hunk 3 above)
Comment (p-lambert) [reply pkg/ebpf/c/tracer-ebpf.c:1365]: Agreed. I think we actually want the real status code, so I'll probably fix that instead of renaming this

File: pkg/network/socket_filter.go (hunk 6 above)
Comment (p-lambert) [reply pkg/network/socket_filter.go:28]: Agreed 👍 I don't mind changing that in this PR

File: pkg/ebpf/c/tracer-ebpf.c (hunk 3 above)
Comment (brycekahle) [reply pkg/ebpf/c/tracer-ebpf.c:1365]: If we want to keep consistent naming with the standard, these are called "status codes".

File: pkg/network/socket_filter.go (hunk 6 above)
Comment (p-lambert) [reply pkg/network/socket_filter.go]: Done

File: pkg/network/ebpf/c/prebuilt/tracer.c [hunk 11]
```diff
@@ -592,16 +642,58 @@ int kprobe__tcp_sendmsg__pre_4_1_0(struct pt_regs* ctx) {
...
 int kretprobe__tcp_sendmsg(struct pt_regs* ctx) {
+#if DEBUG == 1
     int ret = PT_REGS_RC(ctx);
 
     log_debug("kretprobe/tcp_sendmsg: return: %d\n", ret);
+
     // If ret < 0 it means an error occurred but we still counted the bytes as being sent
     // let's increment our miscount count
     if (ret < 0) {
         increment_telemetry_count(tcp_sent_miscounts);
```
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:693]: Why is this being put into a debug-only block?

File: pkg/network/ebpf/c/prebuilt/tracer.c [hunk 12]
```diff
@@ -1158,52 +1250,300 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
...
-        l4_proto = load_byte(skb, ETH_HLEN + offsetof(struct ipv6hdr, nexthdr));
+        l4_proto = load_byte(skb, info->data_off + offsetof(struct iphdr, protocol));
+
+        info->tup.metadata |= CONN_V4;
+        info->tup.saddr_l = load_word(skb, info->data_off + offsetof(struct iphdr, saddr));
+        info->tup.daddr_l = load_word(skb, info->data_off + offsetof(struct iphdr, daddr));
+        info->data_off += sizeof(struct iphdr); // TODO: this assumes there are no IP options
         break;
     default:
+        // TODO: add case for ETH_P_IPV6
```
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:1269]: You still have a TODO here for IPv6

File: pkg/network/ebpf/c/prebuilt/tracer.c [hunk 13]
```diff
@@ -1158,52 +1250,300 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
...
+    // Among other things, but nothing really worked on Kernel 4.4
+    // It seems that indeed support for this type of access by the verifier was added later on:
+    // https://patchwork.ozlabs.org/project/netdev/patch/1475074472-23538-1-git-send-email-jbacik@fb.com/
+    //
+    // What is unfortunate about this is not only that enqueing a HTTP transaction is O(HTTP_BATCH_SIZE),
+    // but also that we can't really increase the batch/page size at the moment because that blows up the eBPF *program* size
+#pragma unroll
+    for (int i = 0; i < HTTP_BATCH_SIZE; i++) {
+        if (i == batch_state->pos) {
+            __builtin_memcpy(&batch->txs[i], http, sizeof(http_transaction_t));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:1397]: Did you want to `break` after the `memcpy`?

File: pkg/network/ebpf/c/prebuilt/tracer.c [hunk 14]
```diff
@@ -1158,52 +1250,300 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
...
+    // Code below is a bit oddly structured in order to make kernel 4.4 verifier happy
+    __u16 status_code = 0;
+    __u8 space_found = 0;
+#pragma unroll
+    for (int i = 0; i < HTTP_BUFFER_SIZE-1; i++) {
+        if (!space_found && buffer[i] == ' ') {
+            space_found = 1;
+        } else if (space_found && status_code < 100) {
+            status_code = status_code*10 + (buffer[i]-'0');
+        }
```
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:1438]: Did you want to `break` once `space_found && status_code >= 100`?

File: pkg/network/ebpf/c/prebuilt/tracer.c [hunk 15]
```diff
@@ -1158,52 +1250,300 @@ int kretprobe__sys_socket(struct pt_regs* ctx) {
...
+        packet_type = HTTP_REQUESTING_GET;
+    } else if ((p[0] == 'P') && (p[1] == 'O') && (p[2] == 'S') && (p[3] == 'T')) {
+        packet_type = HTTP_REQUESTING_POST;
+    } else if ((p[0] == 'P') && (p[1] == 'U') && (p[2] == 'T')) {
+        packet_type = HTTP_REQUESTING_PUT;
+    } else if ((p[0] == 'D') && (p[1] == 'E') && (p[2] == 'L') && (p[3] == 'E') && (p[4] == 'T') && (p[5] == 'E')) {
+        packet_type = HTTP_REQUESTING_DELETE;
+    } else if ((p[0] == 'H') && (p[1] == 'E') && (p[2] == 'A') && (p[3] == 'D')) {
+        packet_type = HTTP_REQUESTING_HEAD;
+    }
```
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/tracer.c:1476]: We are missing the `OPTIONS` and `PATCH` methods. `OPTIONS` is quite common because of CORS requests.

File: pkg/network/ebpf/c/tracer.h [hunk 16]
```diff
@@ -56,6 +56,85 @@ typedef struct {
...
+
+typedef enum {
+    HTTP_UNKNOWN           = 0,
+    HTTP_RESPONDING        = 1,
+    HTTP_REQUESTING_GET    = 1 << 1,
+    HTTP_REQUESTING_POST   = 1 << 2,
+    HTTP_REQUESTING_PUT    = 1 << 3,
+    HTTP_REQUESTING_DELETE = 1 << 4,
+    HTTP_REQUESTING_HEAD   = 1 << 5,
+} http_state_t;
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tracer.h:90]: I mentioned this before, but I'll post it again:

This feels confusing, as it is mixing two concerns:

the state of the transaction: unknown->requesting->responding
The request method
It is also referenced in code as both state and packet_type. Perhaps they should be separate things, or at least different masked parts of the byte.

Method also doesn't need to be a bitmask, since it cannot be multiple methods at the same time.

File: pkg/network/http/monitor.go [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	go func() {
+		var (
+			misses int
+			then   = time.Now()
+			hits   = make(map[int]int)
+			report = time.NewTicker(30 * time.Second)
+			key    = new(httpBatchKey)
+		)
```
Comment (brycekahle) [inline pkg/network/http/monitor.go:99]: You should probably `defer report.Stop()` to cleanup the `time.Ticker`.

**Review by brycekahle — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/prebuilt/tracer.c (hunk 13 above)
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1397]: yes, but it doesn't work.
I can't remember now if the problem was with the loop unrolling itself or if the it was something with the verifier

File: pkg/network/ebpf/c/prebuilt/tracer.c (hunk 14 above)
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1438]: Same as above

File: pkg/network/ebpf/c/prebuilt/tracer.c (hunk 11 above)
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:693]: this `kretprobe` was being used for debug purposes before. now that this is being shared with the http program, I wrapped it in debug block to minimize the overhead when http monitoring is on but debug is off.

File: pkg/network/ebpf/c/prebuilt/tracer.c (hunk 15 above)
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1489]: Done

File: pkg/network/ebpf/c/tracer.h (hunk 16 above)
Comment (p-lambert) [reply pkg/network/ebpf/c/tracer.h]: Done. Let me know if the code now is a bit more clear

File: pkg/network/ebpf/c/prebuilt/tracer.c (hunk 12 above)
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1269]: I think it's fine to merge it without IPv6 support for now

Comment (brycekahle) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1269]: This would be a regression because DNS supports IPv6 right now.

Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1269]: Oh, that's a very good point. I forgot this code path is shared 😳

**Review by olivielpeau — APPROVED** [review]
`Datadog/agent-core` seems to be set as a codeowner of all the files in the PR, although it should only own `pkg/config/config.go`, not sure why as the CODEOWNERS file seems to be up-to-date...

In any case, approving for agent-core, only relevant change for agent-core is in `pkg/config/config.go`.

Please make sure the gitlab pipeline passes before merging. The `arm64` system-probe build is now fixed on master so please pick up the latest changes in master in your branch.

File: pkg/network/ebpf/c/prebuilt/tracer.c (hunk 12 above)
Comment (p-lambert) [reply pkg/network/ebpf/c/prebuilt/tracer.c:1269]: I opened a PR with an integration test for DNS on IPv6 to make sure there won't be any regressions: https://github.com/DataDog/datadog-agent/pull/6944

**Review by sunhay — APPROVED** [review]
👍 for the /pkg/process config changes

**Review by brycekahle — APPROVED** [review]
